cloudinary-image-management-and-manipulation-in-the-cloud-cdn/php/media/class-filter.php

```diff
@@ -522,17 +522,102 @@ public function filter_video_embeds( $html, $id, $attachment ) {
 		$context = $request->get_param( 'context' );
-		if ( 'edit' === $context ) {
-			$data                   = $response->get_data();
-			$content                = $data['content']['raw'];
-			$data['content']['raw'] = $this->filter_out_local( $content );
 
-			$response->set_data( $data );
+		if ( 'edit' !== $context ) {
+			return $response;
 		}
 
+		$data    = $response->get_data();
+		$content = $this->classic_image_to_block( $data['content']['raw'] );
+		$content = $this->video_shortcode_to_block( $content );
+
+		$data['content']['raw'] = $this->filter_out_local( $content );
+		$response->set_data( $data );
+
 		return $response;
 	}
 
+	/**
+	 * Converts video shortcode to a Gutenberg block.
+	 *
+	 * @param string $content
+	 * 
+	 * @return string
+	 */
+	protected function video_shortcode_to_block( $content ) {
+		if ( ! has_shortcode( $content, 'video' ) ) {
+			return $content;
+		}
+
+		$shortcodes = $attrs = array();
+
+		preg_match_all( '/' . get_shortcode_regex( array( 'video' ) ) . '/', $content, $shortcodes );
+
+		for ( $i = 0; $i < count( $shortcodes[3] ); $i++ ) {
+			preg_match_all( '/(mp4|id)="([^"]*)"/i', $shortcodes[3][$i], $attrs );
+	
+			// @TODO: Find a way to generate a video block
+			$video = '<!-- wp:video {"id":' . $attrs[2][1] . '} --><figure class="wp-block-video"><video autoplay src="' . $attrs[2][0] . '"></video></figure><!-- /wp:video -->';
+	
+			$content = str_replace( array( '<p>' . $shortcodes[0][$i] . '</p>', $shortcodes[0][$i] ), $video, $content );
+		}
+
+		return $content;
+	}
+
+	/**
+	 * Converts classic image to a Gutenberg block.
+	 *
+	 * @param string $content
+	 * 
+	 * @return string
+	 */
+	protected function classic_image_to_block( $content ) {
+		if ( strpos( $content, '<img class=' ) === false ) {
+			return $content;
+		}
+
+		$dom = new \DOMDocument();       
+		
+		// Suppress errors here as we are not supplying a full-fledged
+		// HTML document, which will cause loadHTML to throw errors.
+		@$dom->loadHTML( $content, LIBXML_HTML_NODEFDTD );
+		$imgs = $dom->getElementsByTagName( 'img' );
+
+		foreach ($imgs as $img) {
+			// Ensure we're dealing with a proper <img> tag
+			if ( false === strpos( $img->getAttribute( 'class' ), 'wp-image-' ) ) {
+				continue;
+			}
+
+			$img_id = preg_replace( '/[^0-9]/', '', $img->getAttribute( 'class' ) );
+			
+			// Prep the <img> node
+			$img_clone = $img->cloneNode( true );
+			$img_clone->removeAttribute( 'width' );
+			$img_clone->removeAttribute( 'height' );
+			$img_clone->setAttribute( 'class', 'wp-image-' . $img_id );
+
+			// Prep the <figure> node
+			$img_fig = $dom->createElement( 'figure' );
+			$img_fig->appendChild( $img_clone );
+			$img_fig->setAttribute( 'class', 'wp-block-image' );
+			
+			// Surround <figure> with block comments
+			$fragment = $dom->createDocumentFragment();
+			$fragment->appendChild( $dom->createComment( ' wp:image {"id":' . $img_id . ',"sizeSlug":"large"} ' ) );
```

Comment:
@dugajean 

We should try to read the image size from the `$img` element — from the class name, eg. `size-thumbnail` —, and not assuming that we'll always use the `large` size.